spec/index.adoc

```diff
@@ -1949,6 +1949,122 @@ lookup_path(["d"],      pruned_tree) = Found "morning"
+
+[#http-gateway-name-resolution]
+=== Canister resolution
+
+The Gateway needs to know the canister id of the canister to talk to, and obtains that information from the hostname as follows:
+
+1. Check that the hostname, taken from the `Host` field of the HTTP request, is of the form `<name>.raw.ic0.app` or `<name>.ic0.app`, or fail.
+
+2. If the `<name>` is in the following table, use the given canister ids:
++
+.Canister hostname resolution
+|===
+|Hostname |Canister id
+|identity | rdmx6-jaaaa-aaaaa-aaadq-cai
+|nns | qoctq-giaaa-aaaaa-aaaea-cai
+|dscvr | h5aet-waaaa-aaaab-qaamq-cai
+|personhood | g3wsl-eqaaa-aaaan-aaaaa-cai
+
+3. Else, if `<name>` a valid textual encoding of a principal, use that principal as the canister ids.
+
+4. Else fail.
+
+If the hostname was of the form `<name>.ic0.app`, it is a _safe_ hostname; if it was of the form `<name>.raw.ic0.app` it is a _raw_ hostname.
+
+=== Request encoding
+
+The HTTP request is encoded into the `HttpRequest` candid structure.
+
+The `method` field contains the HTTP methd (e.g. `HTTP`), in upper case.
+
+The `url` field contains the URL from the HTTP request line, i.e. without protocol or hostname, and including query parameters.
+
+The `headers` field contains the headers of the HTTP request.
+
+The `body` field contains the body of the HTTP request (without any content encodings processed by the Gateway).
+
+=== Upgrade to update calls
+
+If the canister sets `update = opt true` in the `HttpResponse` reply from `http_request`, then the Gateway ignores all other fields of the reply. The Gateway performs an _update_ call, with the same argument, to `http_request_update`, and uses that response instead.
+
+The value of the `update` field returned from `http_request_update` is ignored.
+
+=== Response decoding and streaming
+
+The Gateway assembles the HTTP response from the given `HttpResponse` record.
+
+The HTTP response status code is taken from the `status_code` field.
+
+The HTTP response headers are taken from the `headers` field.
+
+The HTTP response body is initialized with the value of the `body` field.
+
+If the `streaming_strategy` field is set, the HTTP Gateway then uses further query calls to obtain further chunks to append to the body:
+
+1. If the function reference in the `callback` field of the `streaming_strategy` is not a method of the given canister, the Gateway fails the request.
+
+2. Else, it makes a query call to the give method, passing the `token` value given in the `streaming_strategy` as the argument.
+
+3. That query method returns a `StreamingCallbackHttpResponse`. The `body` therein is appended to the body of the HTTP response. This is repeated as long as the method returns some token in the `token` field, until that field is `null`.
```

Comment:
Slightly off-topic, but I've always been confused exactly how this streaming actually gets implemented. Is it HTTP/1.1 chunked transfer encoding or HTTP/2 streaming?

Should the protocol described in this document talk about whether HTTP/1.1 or HTTP/2 (or both) are supported?

Reply:
This streaming is _only_ concerned with getting the data from the canister to the Gateway, to overcome the message size limit, so it's independent of any HTTP-level streaming. Note that with certification (as it is specified now), the Gateway can't even stream out the data until it has received it all. In the uncertified case it could stream it out (on the TCP level, or the HTTP level) - I don't care and leave that as as implementation choice. Note that if the Gateway is the browser itself it also doesn't make sense to stream.

But it's a good point and I'll clarify!